[FLINK-3187] Introduce RestartStrategy to decouple restarting behaviour from ExecutionGraph

tillrohrmann opened the pull request:
Disclaimer: This PR is based on PR #1468.

## Description

This PR decouples the restart behaviour from the `ExecutionGraph` by introducing the strategy pattern `RestartStrategy`. The `RestartStrategy` encapsulates what will be done in case of a restart.

Currently, the following two `RestartStrategies` are supported:

* `FixedDelayRestartStrategy`: Tries to restart the job a fixed number times with a fixed waiting time in between. This constitutes the ist state.
* `NoRestartStrategy`: No restart, direct failing of job

Having such a decoupling allows us also to set a restart strategy on a per job basis. This renders the necessity to restart the cluster in case of changing the restart delay obsolete. Furthermore, different restart strategies can be used for concurrently running jobs.

Additionally, it is still possible to define a default restart strategy for the cluster. This default strategy is used when there is no other `RestartStrategy` defined for a submitted job.

## API Changes

A `RestartStrategy` can be set using the `setRestartStrategy` method of `(Stream)ExecutionEnvironment`. It looks as follows:

```
ExecutionEnvironment env = ...
env.setRestartStrategy(RestartStrategies.fixedDelay(
  3, // number retry attempts
  1000 // retry delay in milliseconds
));

env.setRestartStrategy(RestartStrategies.noRestart());
```

The default restart strategy is configured using the `restart-strategy` configuration parameter. Depending on the `RestartStrategy` several other configuration parameters can be set. At the moment only the `FixedDelayRestartStrategy` takes more parameters. Those are `restart-strategy.fixed-delay.attempts` and `restart-strategy.fixed-delay.delay`.

In order to configure the `FixedDelayRestartStrategy` as the default strategy, insert the following into the `flink-conf.yaml`.

```
restart-strategy: fixed-delay
restart-strategy.fixed-delay.attempts: 3
restart-strategy.fixed-delay.delay: 10 s
```

 This PR removes the old configuration parameters `execution-retries.default` and `execution-retries.delay` and is thus **API-breaking**.

rmetzger commented:
I like this change because the current restart behavior is not flexible enough. Resolving this issue should be really easy with your change in: https://issues.apache.org/jira/browse/FLINK-3190

With this change, users will execute code in the JobManager and the'll get direct access to the ExecutionGraph. I'm not sure if we should maybe forbid full access to the EC by using a separate interface the strategies can operate on?

tillrohrmann commented:
Thanks for the review @rmetzger.

I think this is not a problem, because the user cannot define restart strategies. In order to set a `RestartStrategy`, the user has to provide a `RestartStrategyConfiguration`. The `RestartStrategyConfiguration` cannot be extended outside the `RestartStrategies` class so that the user cannot define his own `RestartStrategyConfiguration`. Additionally, the strategy itself will only be instantiated from this configuration on the `JobManager` via the `RestartStrategyFactory`. This is also code which cannot be changed by the user via the API.

rmetzger commented:
Whats the status of this PR?

tillrohrmann commented:
Haven't received a +1 for merging yet.

uce commented on flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:
```diff
@@ -237,53 +236,26 @@ public ExecutionConfig setParallelism(int parallelism) {
-	 * of {@code -1} indicates that the system default value (as defined in the configuration)
-	 * should be used.
+	 * Sets the restart strategy configuration which defines which restart strategy shall be used
+	 * for the execution graph of the corresponding job.
```
I would add a `<code></code>` example showing the `RestartStrategies`, which will be the common way to configure it I guess.

The text could maybe also be simplified at the end by removing execution graph and corresponding job. The average user will not know what it is. On the other hand, it might be a good pointer for someone who wants to work on it.

tillrohrmann replied:
Agreed. Good point. I've simplified the the description and added a code example.

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/restart/FixedDelayRestartStrategy.java:
```diff
@@ -0,0 +1,113 @@
+					// do the delay
+					Thread.sleep(delayBetweenRestartAttempts);
+				}
+				catch(InterruptedException e){
```
whitespace after catch

tillrohrmann replied:
good catch

uce commented on docs/apis/fault_tolerance.md:
```diff
@@ -193,73 +193,169 @@ state updates) of Flink coupled with bundled sinks:
-----------------------------------------------
+Flink supports different restart strategies which control how the jobs are restarted in case of a failure.
+The cluster can be started with a default restart strategy which is always used when no job specific restart strategy has been defined.
+In case that the job is submitted with a restart strategy, this strategy overrides the cluster's default setting.
```
Should we add a sentence that the current cluster default is `no restart`?

tillrohrmann replied:
Yes, definitely

uce commented on flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java:
```diff
@@ -0,0 +1,116 @@
+
+		/**
+		 * Returns a description which is shown in the web interface
+		 * @return
```
Javadocs

tillrohrmann replied:
Added them

uce commented:
Changes look very good. I've tried it out with the two provided `RestartStrategies` and it works like a charm. ;) (I didn't try to write a custom strategy, but that is OK I guess. The code looks straight forward there.)

I had some trivial inline comments, which should not block merging this if you don't agree. +1 to merge.

tillrohrmann commented:
I addressed Ufuk's comments and rebased on the latest master + PR #1577. If we are ok with the changes, which are API breaking because they change how the restart behaviour is defined, then I would  like to merge this PR.

Any objections?

tillrohrmann commented:
If nobody objects, then I would like to merge this PR, since it will give us more flexibility in the future with respect to restarting strategies.

uce commented:
Thanks for addressing the comments. Changes look good!

Regarding the API breaking config key changes: is it feasible to just deprecate the old configuration keys instead of removing them? If set, we can just translate it to the matching restart strategy.

tillrohrmann commented:
Alright, I've reintroduced the old execution attempts and delay configuration values and the API calls at the `ExecutionEnvironment`.

The behaviour is now the following: 

1. If an explicit `RestartStrategy` is set for a job, it is taken. 

2. Otherwise it is checked whether the number of retries and retry delay has been set at the `ExecutionEnvironment`/`ExecutionConfig`. If this is the case, then a `FixedDelayRestartStrategy` is instantiated with these values.  

3. If no explicit `RestartStrategy` has been defined for the job, then the default restart strategy of the `JobManager` is used. The default restart strategy is defined the following way:

3.1. If the configuration contains a configuration value `restart-strategy`, then this defines the used `RestartStrategy`.

3.2. If `restart-strategy` is not set, then the old `execution-retries.default` and `execution-retries.delay` configuration values are checked. If they are set with `execution-retries.default > 0` and `execution-retries.delay >= 0`, then a `FixedDelayRestartStrategy` is instantiated with the respective values. This is then used as the default restart strategy. If these values are not defined, then a `NoRestartStrategy` is instantiated.

This should be not API breaking unless people used the `setExecutionRetries` at the `JobGraph` or the `Plan`.

uce commented:
Thank you very much! Behaviour is as I would expect it.

Last minor comment: can you add a comment in the deprecated methods (both Java and Scala) with a reference to the restart strategy while merging? I think this will make the transition easier for users.

+1 to merge.

tillrohrmann commented:
Sure, will do. Thanks for the review :-)

tillrohrmann commented:
Merging this.
